File: src/1_100/0030_substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if(s==None or len(s)==0 or words==None or len(words)==0):
            return []
        word_dict = dict()
        for w in words:
            if(word_dict.has_key(w)):
                word_dict[w] += 1
            else:
                word_dict[w] = 1
        word_arr_size = len(words) * len(words[0])
        single_word_size = len(words[0])
        res = []
        for i in range(len(s)-word_arr_size+1):
            tmp_str= s[i:i+word_arr_size]
            d = dict(word_dict)
            for j in range(0,len(tmp_str),single_word_size):
                key = tmp_str[j:j+single_word_size]
                if(d.has_key(key)):
                    if(d[key] <= 0):
                        break
                    else:
                        d[key] -= 1
                        if(d[key] == 0):
                            del d[key]
                else:
                    break
            if(len(d)==0):
                res.append(i)    
        return res

File: src/1_100/0030_substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.java
class Solution {
	public List<Integer> findSubstring(String s, String[] words) {
		if(s==null || "".equals(s) || words==null || words.length==0) {
			return new ArrayList<Integer>();
		}
		//将每个单词以及出现的频率记录到map中
		Map<String,Integer> words_map = new HashMap<String,Integer>();
		for(String str : words) {
			if(words_map.containsKey(str)) {
				words_map.put(str,words_map.get(str)+1);
			} else {
				words_map.put(str,1);
			}
		}
		List<Integer> res = new ArrayList<Integer>();
		//words中一个单词的长度，以及words的总长度
		int one_word_size = words[0].length();
		int all_word_size = words.length * one_word_size;
		//遍历整个字符串，注意循环的结束条件
		for(int i=0;i<s.length()-all_word_size+1;i++) {
			//每次取 all_word_size长度的子串
			String tmp = s.substring(i,i+all_word_size);
			HashMap<String,Integer> d = new HashMap<String,Integer>(words_map);
			//将子串和临时map进行比较
			for(int j=0;j<tmp.length();j+=one_word_size) {
				//从子串tmp中取出one_word_size长度的子串，看是否出现在临时map中
				//如果是就将临时map记录的频率-1，如果不在就跳出循环
				String key = tmp.substring(j,j+one_word_size);                
				if(d.containsKey(key)) {
					d.put(key,d.get(key)-1);
					if(d.get(key)==0) {
						d.remove(key);
					}
				} else {
					break;
				}
			}
			//当内层循环遍历完后，如果临时map为空则表示全部匹配上了
			//记录数组的下标
			if(d.size()==0) {
				res.add(i);
			}
		}
		return res;
	}
}